pkg/sql/execinfra/flow_context.go

```diff
@@ -114,6 +122,40 @@ func (ctx *FlowCtx) Codec() keys.SQLCodec {
 	return ctx.EvalCtx.Codec
 }
 
+// TableDescriptor returns a catalog.TableDescriptor object for the given
+// descriptor proto, using the descriptors collection if it is available.
+func (ctx *FlowCtx) TableDescriptor(desc *descpb.TableDescriptor) catalog.TableDescriptor {
+	if desc == nil {
+		return nil
+	}
+	if ctx != nil && ctx.Descriptors != nil {
+		leased, _ := ctx.Descriptors.GetLeasedTableByID(ctx.EvalCtx.Ctx(), ctx.Txn, desc.ID)
+		if leased != nil && leased.GetVersion() == desc.Version {
```

Comment:
Is checking the version even necessary at all?

Reply:
I believe that it is, yes. Or, at least, I certainly don't want to have to convince myself it's not. The alternative is that the descriptor is either newer or older. 

It's worth noting that much hand-wringing was done to convince ourselves that it's safe to use enums that are out of sync because the values are always there if they're going to be used. That made me uneasy at the time.  I worry about more complex cross-descriptor interactions. 

My preference in the longer term is to encode id and version pairs for all needed descriptors and then make it cheap to retrieve them in one go. Generally they'll be cached. That way we'd have nice invariants that all of the exact descriptors matched across machines. 

The edge case is if the descriptor hasn't been committed yet. In that case we'll still need to serialize it all. That fact has always made me sort of sad.

Reply:
Thanks for this information, I was also wondering about the enums. I agree with your general assessment.